internal/infrastructure/ceph/rbd.go

```diff
@@ -267,6 +357,304 @@ func applyDiffDataRecords(
+
+				// Compute the portion of this chunk that should be written.
+				chunkStartPos := chunkIndex * rawChecksumChunkSize
+				chunkEndPos := chunkStartPos + rawChecksumChunkSize
+				writeStartPos := max(offset, chunkStartPos)
+				writeEndPos := min(offset+length, chunkEndPos)
+				writeSize := writeEndPos - writeStartPos
+
+				if writeSize == rawChecksumChunkSize {
+					if err := processEntireChunkWrite(
+						rawFile,
+						rawChecksumFile,
+						diffFileReader,
+						tag,
+						chunkIndex,
+						chunkStartPos,
+						rawChecksumChunkSize,
+						csumZero,
+					); err != nil {
+						return err
+					}
+				} else {
+					if err := processPartialChunkWrite(
+						rawFile,
+						rawChecksumFile,
+						diffFileReader,
+						tag,
+						chunkIndex,
+						chunkStartPos,
+						writeStartPos,
+						writeSize,
+						rawChecksumChunkSize,
+					); err != nil {
+						return err
+					}
+				}
+			}
+
+		case 'e': // END
+			return nil
+
+		default:
+			return fmt.Errorf("unexpected tag: %c", tag)
+		}
+	}
+}
+
+func processEntireChunkWrite(
+	dstFile *os.File,
+	dstChecksumFile *os.File,
+	diffFileReader *bufio.Reader,
+	tag byte,
+	chunkIndex uint64,
+	chunkStartPos uint64,
+	rawChecksumChunkSize uint64,
+	csumZero uint64,
+) error {
+	var csumUpdated uint64
+	chunkData := make([]byte, rawChecksumChunkSize)
+
+	if tag == 'w' { // UPDATED DATA
+		_, err := io.ReadFull(diffFileReader, chunkData)
+		if err != nil {
+			if errors.Is(err, csumreader.ErrChecksumMismatch) {
+				return csumreader.ErrChecksumMismatch
+			}
+			return fmt.Errorf("failed to read chunk data: %w", err)
+		}
+
+		csumUpdated = xxhash.Sum64(chunkData)
+
+		// Update checksum file
+		if err := writeChecksumAtIndex(dstChecksumFile, chunkIndex, csumUpdated); err != nil {
+			return err
+		}
+
+		// Update raw.img
+		if _, err := dstFile.Seek(int64(chunkStartPos), io.SeekStart); err != nil {
+			return fmt.Errorf("failed to seek in destination file: %w", err)
+		}
+		if _, err := dstFile.Write(chunkData); err != nil {
+			return fmt.Errorf("failed to write to destination file: %w", err)
+		}
+	} else { // ZERO DATA
+		// Update checksum file
+		if err := writeChecksumAtIndex(dstChecksumFile, chunkIndex, csumZero); err != nil {
+			return err
+		}
+
+		// Write zeros using fallocate
+		if err := unix.Fallocate(
+			int(dstFile.Fd()),
+			unix.FALLOC_FL_KEEP_SIZE|unix.FALLOC_FL_PUNCH_HOLE,
+			int64(chunkStartPos),
+			int64(rawChecksumChunkSize),
+		); err != nil {
+			return fmt.Errorf("failed to write zero data: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func processPartialChunkWrite(
+	rawFile *os.File,
+	rawChecksumFile *os.File,
+	diffFileReader *bufio.Reader,
+	tag byte,
+	chunkIndex uint64,
+	chunkStartPos uint64,
+	writeStartPos uint64,
+	writeSize uint64,
+	rawChecksumChunkSize uint64,
+) error {
+	// Read the raw image checksum and its current chunk contents.
+	storedCsum, err := readChecksumAtIndex(rawChecksumFile, chunkIndex)
```

Comment:
It's better to use consistent naming scheme for variables for checksum .

```suggestion
	csumStored, err := readChecksumAtIndex(rawChecksumFile, chunkIndex)
```

Of course, `storedCsum`, `currentCsum`, `updatedCsum` is also OK. It's up to you.